src/errors.rs

```diff
@@ -35,64 +35,55 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
-use std::{io, ffi, num};
 use std::fmt::{self, Formatter, Display};
+use std::error::Error;
+use TracerState;
 
 #[derive(Debug)]
 pub enum HWTracerError {
-    // Wrapped errors from elsewhere.
-    FFIIntoString(ffi::IntoStringError),
-    FFINul(ffi::NulError),
-    IO(io::Error),
-    NumParseInt(num::ParseIntError),
-    // Our own errors.
-    CFailure,
-    HardwareSupport(String),
-    InvalidFileName(String),
-    TracerAlreadyStarted,
-    TracerDestroyed,
-    TracerNotStarted,
-    TracingNotPermitted(String),
-}
-
-impl From<ffi::IntoStringError> for HWTracerError {
-    fn from(err: ffi::IntoStringError) -> Self {
-        HWTracerError::FFIIntoString(err)
-    }
-}
+    // Non-fatal errors.
+    HWBufferOverflow,     // The trace buffer being used by the hardware overflowed.
+                          // This is considered a non-fatal error since retrying the tracing may
+                          // succeed.
+    HWSupport(String),    // The hardware doesn't support a required feature. Not fatal for the
+                          // same reason as `TracingPerms`.
+    TracingPerms(String), // Tracing is not permitted. Non-fatal as the consumer could carry on
```

Comment:
Not a big fan of this name: `PermissionDenied` or something might be clearer?

Reply:
Do you intend to make it a more general error? Otherwise we should qualify it with something that makes it's specific to tracing capabilities. e.g. `TraceingPermissionsDenied`, which I find too long.

Reply:
Since everything here relates to tracing, I don't see how that prefix is useful TBH.

Reply:
OK, let's go for `Permissions`.